### PR TITLE
Improved docker image generation/configuration [Upstream Retarget: Pending Review]

### PR DESCRIPTION
**Pull request by bengreenier**
This adds support for enhanced docker image generation using the tooling built into the core project. It's based off the discoveries made in https://github.com/bengreenier/sgx-lkl-turtles by myself, and my team.

We've made these changes to support a development scenario that we think is quite promising going forward...Where the LKL app is defined in a self-contained (and `alpine` based) Dockerfile, that some tooling parses and processes to produce an image containing the lkl runtime (`sgx-lkl-run`) and the `img`, along with the desired runtime arguments and environment variables. Hopefully this pattern is found to be beneficial to other folks as well!

This doesn't change the shell API, but does enhance it's capabilities in the following ways:
+ Accepts container tags that aren't running/haven't run
+ Accepts container tags with `:` on oses that don't support `:` in fs paths (replaces with `_`)
+ Reads `WORKDIR` from docker image, and maps it at runtime
+ Reads `ENTRYPOINT` from docker image, and maps it at runtime
+ Reads `ENV` from docker image, and maps it at runtime
+ Overrides `ENTRYPOINT` with command line args (shell API args take precedence)
+ Exits (`exit 1`) when neither a docker `ENTRYPOINT` nor shell API arg are given (we must have an entrypoint)
+ Clarifies output tag via echo message (`{inputTag}-secure`)
+ Clarifies runtime requirement for `--privileged` via echo message
+ Reduces stdout noise in final container (no more `net.ipv4.ip_forward=1` message)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -55,7 +55,7 @@ WORKDIR /sgx-lkl
 
 RUN apt-get update && apt-get install -y \
   sudo \
-  iproute2 iptables net-tools \
+  iproute2 iptables net-tools libjson-c-dev libprotobuf-c-dev \
```
Comment (bengreenier) [inline Dockerfile]: @bengreenier test if these are still needed, given we had other runtime issues present when we added this as a fix

File: Dockerfile [hunk 2]
```diff
@@ -80,6 +82,8 @@ CMD ["/bin/bash", "-c", "sudo ip tuntap add dev sgxlkl_tap0 mode tap user user \
     && sudo iptables -I FORWARD -m state -s 10.0.1.0/24 --state NEW,RELATED,ESTABLISHED -j ACCEPT \
     && sudo iptables -I FORWARD -m state -d 10.0.1.0/24 --state NEW,RELATED,ESTABLISHED -j ACCEPT \
     && sudo iptables -t nat -A POSTROUTING -s 10.0.1.0/24 ! -d 10.0.1.0/24 -j MASQUERADE \
-    && sudo sysctl -w net.ipv4.ip_forward=1 \
+    && sudo sysctl -w net.ipv4.ip_forward=1 > /dev/null \
```
Comment (bengreenier) [inline Dockerfile]: this is just for quality of life - cleans up the docker run stdout

File: Dockerfile [hunk 3]
```diff
@@ -80,6 +82,8 @@ CMD ["/bin/bash", "-c", "sudo ip tuntap add dev sgxlkl_tap0 mode tap user user \
...
     && sudo chown user /dev/net/tun \
-    && SGXLKL_TAP=sgxlkl_tap0 SGXLKL_HEAP=2500M /sgx-lkl/build/sgx-lkl-run /sgx-lkl/enclave_rootfs.img ${env_binary_cmd} ${env_binary_args}"]
+    && export SGXLKL_HEAP=2500M \
```
Comment (bengreenier) [inline Dockerfile]: we could port in the configurability here - if we think that's useful

File: sgx-lkl-docker.sh [hunk 4]
```diff
@@ -5,6 +5,10 @@
 # docker-machine create --driver generic --generic-ip-address=<host> --generic-ssh-key ~/.ssh/<key> --generic-ssh-user=<user> <machine_id>
 
 set -e
+if  [ "${OSTYPE}" == "msys" ]; then
```
Comment (bengreenier) [inline sgx-lkl-docker.sh]: @bengreenier  this is the only windows fix needed - we should pull this out to a windows fix branch and remove the other related windows-fix bits from this PR

File: sgx-lkl-docker.sh [hunk 5]
```diff
@@ -65,18 +69,24 @@ function build() {
...
+
     if [ "$(uname)" == "Darwin" ]; then
         echo Note that SSH agent access inside the Docker container under macOS depends on https://github.com/uber-common/docker-ssh-agent-forward
         SSH_AGENT_WORKAROUND=`pinata-ssh-mount`
     elif [ "$(uname)" == "Linux" ]; then
         SSH_AGENT_WORKAROUND="--volume $SSH_AUTH_SOCK:/ssh-agent --env SSH_AUTH_SOCK=/ssh-agent"
+    elif [ "${OSTYPE}" == "msys" ]; then
+        # From: https://stackoverflow.com/a/13701495
+        VOLUME_MOUNTS=$(echo "${VOLUME_MOUNTS}" |  sed -e 's/^\///' -e 's/^./\0:/')
+        DOCKER_SHELL='//bin/bash'
```
Comment (bengreenier) [inline sgx-lkl-docker.sh]: @bengreenier this is one of the "other windows fixes" that can go

File: Dockerfile (hunk 3 above)
Comment (bengreenier) [reply Dockerfile]: ah this was me being sleepy. this is just a default, override it to your hearts content

File: deploy_container_test.sh [hunk 6]
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/env bash
+
+#
+# This script helps to validate the sgx-lkl-docker behavior for the deploy-container scenario
```
Comment (bengreenier) [inline deploy_container_test.sh]: probably won't check this in, will have in HEAD~1 and then HEAD will be without when we PR.

That way - you can see how i validated, and even get the file if you want, but not actually have it on disk @ HEAD

File: Dockerfile (hunk 1 above)
Comment (bengreenier) [reply Dockerfile]: they are

File: deploy_container_test.sh (hunk 6 above)
Comment (bengreenier) [reply deploy_container_test.sh]: removed with https://github.com/bengreenier/sgx-lkl/pull/4/commits/9daa692a04ce1854f0d84259074164b9d999edcd

**Comment by bengreenier** [comment]
this is now ready for review/testing by others.

Have verified that test script (removed with https://github.com/bengreenier/sgx-lkl/pull/4/commits/9daa692a04ce1854f0d84259074164b9d999edcd) worked in WIN10 + MSYS-bash (from git-for-windows) with `export MSYS_NO_PATHCONV=1`.

Removing the need for `MSYS_NO_PATHCONV=1` is separately covered by https://github.com/lsds/sgx-lkl/pull/69

**Comment by bengreenier** [comment]
will retarget this PR to upstream after some peers have done further validation

**Comment by bengreenier** [comment]
PS my build-validation is still broken..the failures there don't indicate actual build failures, just process failures.

**Comment by bengreenier** [comment]
> 
> 
> PS my build-validation is still broken..the failures there don't indicate actual build failures, just process failures.

This is no longer the case.

**Comment by bengreenier** [comment]
@jjcollinge can you take a look at this?

File: sgx-lkl-docker.sh [hunk 7]
```diff
@@ -146,43 +146,70 @@ function deploy-container() {
...
+
+    if [[ -z "${SRC_CONTAINER_WD}" ]]; then
+        echo "Container had no WORKDIR. Setting to '/'."
+        SRC_CONTAINER_WD=/
+    fi
+
+    echo "Creating secure container '${SEC_CONTAINER}' from container '${SRC_CONTAINER}' with the following configuration:"
+    echo "Launch Command: ${SRC_CONTAINER_ENTRY}"
+    echo "Launch Working Directory: ${SRC_CONTAINER_WD}"
+    echo "Environment Variables: ${SRC_CONTAINER_ENV}"
```
[the remark this answers is not in the document]
Comment (jjcollinge) [reply sgx-lkl-docker.sh]: I'd leave it for now - let upstream maintainer's decide if it's an issue!

**Comment by jjcollinge** [comment]
LGTM

**Pull request closed without merging**